Step2

rootkwak528 opened the pull request:
아토 짱짱맨 !!!
저 이제 좀 객체를 만들 줄 아는거 같은데 ..
이거 또 아토가 보고 한숨 푹푹 쉬면 어카지 ..

kakao-root-directory committed:
refactor
- BiFunction -> BinaryOperator
- indent depth 1
- enum 파일 분리
- convertOperator() 메서드 enum 안으로
- Pattern 객체 사용
- DisplayName 추가

kakao-root-directory committed:
test : step2 테스트 추가

kakao-root-directory committed:
feat : step2 구현

daum7766 commented on src/main/java/Lotto/Lotto/LottoGame.java:
```diff
@@ -0,0 +1,5 @@
+package Lotto.Lotto;
```
패키지 컨벤션은 모두 소문자입니다.

rootkwak528 replied:
띠용 !!

daum7766 commented on src/main/java/Lotto/Lotto/LottoMoney.java:
```diff
@@ -0,0 +1,21 @@
+public class LottoMoney {
+    private static final Integer TICKET_PRICE = 1000;
+
+    private final Integer buyMoney;
+    private final Integer lottoCount;
```
원시타입이 아니라 래퍼타입을 사용한 이유가 있을까요?

rootkwak528 replied:
흠 딱히 없는거 같아요
원시타입이 더 나은 선택인가요?

daum7766 commented on src/main/java/Lotto/Lotto/LottoResult.java:
```diff
@@ -0,0 +1,44 @@
+        rewardCountMap = new HashMap<>();
+    }
+
+    LottoResult(LottoMoney lottoMoney, LottoTickets lottoTickets, LottoWinNumbers lottoWinNumbers) {
+        this.lottoMoney = lottoMoney;
+        rewardCountMap = new HashMap<>();
+        for (LottoTicket lottoTicket: lottoTickets.getLottoTickets()) {
+            LottoReward lottoReward = lottoTicket.match(lottoWinNumbers);
+            this.increaseCount(lottoReward);
+        }
+    }
```
생성자에서는 비즈니스 로직이 안들어가는게(for문 처리 등) 좋습니다.
정적팩터리 메서드를 도입해보는건 어떨까요?

daum7766 commented on src/main/java/Lotto/Lotto/LottoResult.java:
```diff
@@ -0,0 +1,44 @@
+
+public class LottoResult {
+
+    private final Map<LottoReward, Integer> rewardCountMap;
```
생성자에서 new하지않고 변수에 바로 new 로해서 생성하면 좋을 것 같아요.

daum7766 commented on src/main/java/Lotto/Lotto/LottoResult.java:
```diff
@@ -0,0 +1,44 @@
+
+    public Float profit() {
+        float totalRewardMoney = 0f;
+        for (Map.Entry<LottoReward, Integer> entry: rewardCountMap.entrySet()) {
+            totalRewardMoney += entry.getValue() * entry.getKey().getRewardMoney();
+        }
```
👍 for each 좋네요.

daum7766 commented on src/main/java/Lotto/Lotto/LottoResult.java:
```diff
@@ -0,0 +1,44 @@
+        for (Map.Entry<LottoReward, Integer> entry: rewardCountMap.entrySet()) {
+            totalRewardMoney += entry.getValue() * entry.getKey().getRewardMoney();
+        }
+        return totalRewardMoney / lottoMoney.getBuyMoney();
```
lottoMoney.getBuyMoney() 가 0일때는 어떻게 동작하나요?

rootkwak528 replied:
인풋에 대한 예외처리는 크게 고려하지 않고 작성했는데 말이죠 .. ; _ ;
사용자 인풋에 대한 검증이 추가는 일부러 제외하고 비즈니스 로직에만 집중해서 구현중이라 (시간 효율을 위해 !)
buyMoney 역시 사용자 인풋이기 때문에 굳이 엣지케이스 처리를 하지 않았습니다 !

daum7766 commented on src/main/java/Lotto/Lotto/LottoReward.java:
```diff
@@ -0,0 +1,34 @@
+        return rewardMoney;
+    }
+
+    public static LottoReward valueOfMatchCount(final Integer matchCount) {
+        return Arrays.stream(LottoReward.values())
+                .filter(value -> value.matchCount.equals(matchCount))
+                .findAny()
+                .orElse(NO);
+    }
```
👍

daum7766 commented on src/main/java/Lotto/Lotto/LottoTicket.java:
```diff
@@ -0,0 +1,34 @@
+        this.numbers = Random.generateRandomLottoNumbers();
+    }
+
+    public LottoTicket(String numbers) {
+        this.numbers = Arrays.stream(numbers.split(SPLITTER))
+                .map(Integer::parseInt)
+                .collect(Collectors.toList());
+    }
```
문자열을 split에서 int로 변환하고 List로 만드는게 LottoTicket의 역할일까요? 루트의 생각이 궁금합니다.

rootkwak528 replied:
흠 해당 변환 로직은 별도 util로 옮기고,
생성자에서는 인자로 List<Integer> numbers를 받는게 나을 거 같네요.

daum7766 commented on src/main/java/Lotto/Lotto/LottoTicket.java:
```diff
@@ -0,0 +1,34 @@
+                .collect(Collectors.toList());
+    }
+
+    public List<Integer> getNumbers() {
+        return new ArrayList<>(numbers);
+    }
```
방어로직 분리 굿!

daum7766 commented on src/main/java/Lotto/Lotto/LottoWinNumbers.java:
```diff
@@ -0,0 +1,22 @@
+
+    final List<Integer> numbers;
+
+    public LottoWinNumbers(String numbers) {
+        this.numbers = Arrays.stream(numbers.split(SPLITTER))
+                .map(str -> Integer.parseInt(str.trim()))
+                .collect(Collectors.toList());
+    }
```
이것도 위에랑 같은 구조인것 같습니다.

daum7766 commented on src/main/java/Lotto/View/InputView.java:
```diff
@@ -0,0 +1,20 @@
+
+    public static LottoWinNumbers inputWinNumbers() {
+        Scanner scanner = new Scanner(System.in);
+        System.out.println("\n지난 주 당첨 번호를 입력해 주세요.");
```
운영체제에 따라 \n 혹은 \n\r 이 개행일 수 있습니다.
System.lineSeparator();를 이용하면 운영체제에 맞는 개행문자를 반환합니다.

daum7766 commented on src/test/java/StringCalculator/View/InputViewTest.java:
```diff
@@ -9,29 +10,37 @@
     public void nullTest() {
         assertThatThrownBy(() -> InputView.validate(null))
-                .isInstanceOf(IllegalArgumentException.class);
+                .isInstanceOf(IllegalArgumentException.class)
```
정확한 에러확인을 위해 isExactlyInstanceOf로 검증하는게 좋습니다.
isInstanceOf를 이용하면 상속받고 있는 클래스는 모두 통과하게 됩니다.

daum7766 commented on src/test/java/Lotto/Lotto/LottoMoneyTest.java:
```diff
@@ -0,0 +1,16 @@
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class LottoMoneyTest {
+    @ParameterizedTest
+    @CsvSource(value = {"10000,10", "14500,14", "900,0"}, delimiter = ',')
+    @DisplayName("돈으로 몇장의 티켓을 사는지 확인")
+    public void buyCountTest(Integer money, Integer ticketCount) {
+        assertThat(new LottoMoney(money).getLottoCount()).isEqualTo(ticketCount);
+    }
```
로또기능과 관련된 테스트가 전체적으로 부족한것 같습니다.

rootkwak528 replied:
힝 ; _ ;

daum7766 reviewed (commented):
나중에 더 확인할게요~ (시간이 부족해서 간단하게 봤습니다)

kakao-root-directory committed:
refactor : 리뷰 반영

by 용주님
- 사용하지 않는 LottoGame 삭제
- LottoMoney에서 LottoCount 클래스 분리
- LottoResult에 lottoMoney 멤버 변수 사용 대신 profit()에 lottoMoney 인자로 주입하는 것으로 대체
- 숫자에 _ 사용
- enum.values()로 순회
- utils 패키지 내 클래스명, 메서드명 수정
- LottoResultTest.java를 위한 생성자 추가
- 당첨번호 중복 검증 메서드 추가

by 유기님
- 패키지 컨벤션 소문자
- 생성자에서 비즈니스 로직 있는 경우 정적팩토리 메서드로 분리
- LottoTicket에서 StringToListInteger 변환로직 별도 클래스로 분리
- \n 대신 System.lineSeparator() 사용
- isInstanceOf 대신 isExactlyInstanceOf 사용
